### DISCUSSION TITLE
AuxVariables indexing for nodes

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#18228] [Q&A General] [answered]
Hello,
I am trying to solve a variable that is being locally updated on each node where it accumulates over time and used in the main kernel in my application. The updating of the variable is very simple and i am trying to avoid coupling nonlinear variables as the problem is complex as it is. So i think that using an AuxKernel is the way to go.
However, I am not sure how the nodal variables are accessed in AuxKernels, since `_u_old` is evaluated at `_qp`, and I am aware also that there is a `_nodal` flag for AuxKernels but I still do not know how it is indexed for each node and didn't find examples for it. As far as I have seen in the code is that `isNodal()` function determines whether `_qp=0` but it is not clear how to index different nodes. Also, are the initial conditions for AuxVariables are set to zero by default?

This is not a boundary variable, so would you recommend using AuxKernel or Material for this?

Thank you,
Khaled

**Comment by jiangwen84** [comment, marked as answer]
If your aux variable is nodal type, e.g. LAGRANGE, the `Auxkerne`l will set the value directly for each node. `_qp=0` refers to the node. You can use `_current_node->id()` to give you the node id. 

It seems to be correct that IC for aux variables are zero by default.

**Reply by KhaledNabilSharafeldin** [reply]
is `_u_old` refer to the nodal values or does it always refer to the qp values. Does it get stored at qp or at nodes for the case of nodal auxVariables.

**Reply by loganharbour** [reply]
For all AuxKernels and Kernels that access Variables and AuxVariables, the size of `_u`, `_u_old`, etc will be set automatically depending on if the kernel is nodal or not. That is, if it is nodal it will have a size of 1 (access nodal value at `[0]`), and if it is not nodal it will have the size of the number of quadrature points on the current element.
